### PR TITLE
Abstracted cloud storage - Amazon/Google/Azure/Local

### LINKED ISSUE #239: Add unified file storage solution
Add in a way to store files within the system

@NicolasDorier  suggested to store them in the db directly but I'm a bit wary of that due to issues that could come with it (more details here: https://softwareengineering.stackexchange.com/questions/150669/is-it-a-bad-practice-to-store-large-files-10-mb-in-a-database)

I propose we use something like this: https://github.com/geeklearningio/gl-dotnet-storage

### PR DESCRIPTION
**Pull request by Kukks**
Currently restricted only to Admins but this can be extended with extra settings to allow non-admin stores or apps to use it to upload css or image files( with a file size restriction)

I also plan to use this to handle the FastSync direct integration into Btcpay

Another use-case is https://github.com/btcpayserver/btcpayserver-docker/issues/116

closes #239

image
image
image
image
image
image

### CODE REVIEW

**Comment by NicolasDorier** [comment]
I am worried that people get confused about it and think that everything get saved into the storage, when it is not the case.

How about, instead of having a menu specifically for it, just setting it up as a "Service"?

Also, I am worried that people use file storage on the docker deploy, because if they do, they will lose the files when the container restarts. So if file storage, the root of the storage should be in the datadir of the chain.

**Comment by Kukks** [comment]
> Also, I am worried that people use file storage on the docker deploy, because if they do, they will lose the files when the container restarts. So if file storage, the root of the storage should be in the datadir of the chain.

Hmm, we need them to be in the wwwroot to be publicly available. Maybe another volume? Or I think there is a FileProvider trick I can do to map the folder to the web if I remember correctly

**Comment by NicolasDorier** [comment]
not the wwwroot,  the wwwroot also lose the data

**Comment by Kukks** [comment]
@NicolasDorier  using data dir now

**Comment by Kukks** [comment]
Moved to services section and added a `/storage/{fileId}` endpoint

**Comment by NicolasDorier** [comment]
tests crash

File: BTCPayServer/Controllers/ServerController.Storage.cs [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+            var savedSettings = await _SettingsRepository.GetSettingAsync<StorageSettings>();
+            if (forceChoice || savedSettings == null)
+            {
+                return View(new StorageSettings()
+                {
+                    Provider = savedSettings?.Provider ?? BTCPayServer.Storage.Models.StorageProvider.FileSystem
+                });
+            }
+
+            return RedirectToAction("StorageProvider", new
```
Comment (NicolasDorier) [inline BTCPayServer/Controllers/ServerController.Storage.cs]: can you use `nameof(StorageProvider)`

File: BTCPayServer/Controllers/ServerController.Storage.cs [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+            return RedirectToAction("StorageProvider", "Server", new
+            {
+                provider = viewModel.Provider.ToString()
+            });
+        }
+
+        [HttpGet("server/storage/{provider}")]
+        public async Task<IActionResult> StorageProvider(string provider)
+        {
+            var storageProvider = Enum.Parse(typeof(StorageProvider), provider);
```
Comment (NicolasDorier) [inline BTCPayServer/Controllers/ServerController.Storage.cs]: uncaught exception  if provider can't be parsed or null

File: BTCPayServer/Controllers/ServerController.Storage.cs [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+            var data = (await _SettingsRepository.GetSettingAsync<StorageSettings>()) ?? new StorageSettings();
+
+            var storageProviderService =
+                _StorageProviderServices.Single(service => service.StorageProvider().Equals(storageProvider));
+
+            var viewName = $"Edit{storageProvider}StorageProvider";
```
Comment (NicolasDorier) [inline BTCPayServer/Controllers/ServerController.Storage.cs]: don't do magic strings, use `nameof()` when you call `View` instead.

File: BTCPayServer/BTCPayServer.csproj [hunk 4]
```diff
@@ -69,7 +69,12 @@
   </ItemGroup>
 
 	<ItemGroup>
-    <PackageReference Include="Microsoft.AspNetCore.App" Version="2.1.9" />
+    <PackageReference Include="Microsoft.AspNetCore.App" Version="2.1.6" />
```
Comment (NicolasDorier) [inline BTCPayServer/BTCPayServer.csproj]: don't touch this version

File: BTCPayServer/Controllers/ServerController.Storage.cs [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+                        Severity = StatusMessageModel.StatusSeverity.Error,
+                        Message = $"{provider} provider is not supported"
+                    }.ToString()
+                });
+            }
+
+            var data = (await _SettingsRepository.GetSettingAsync<StorageSettings>()) ?? new StorageSettings();
+
+            var storageProviderService =
+                _StorageProviderServices.Single(service => service.StorageProvider().Equals(storageProvider));
```
Comment (NicolasDorier) [inline BTCPayServer/Controllers/ServerController.Storage.cs]: if not found, then unhandled exception. Should also be case insensitive.

File: BTCPayServer/BTCPayServer.csproj (hunk 4 above)
Comment (Kukks) [reply BTCPayServer/BTCPayServer.csproj]: must have been a rebase fuckup

**Comment by NicolasDorier** [comment]
Need tests as well.

You should use dev secrets as I did with the transifex thing, so a dev can run manually the tests if he has some amazon creds.

File: BTCPayServer.Tests/BTCPayServer.Tests.csproj [hunk 6]
```diff
@@ -33,4 +33,8 @@
   <ItemGroup>
     <ProjectReference Include="..\BTCPayServer\BTCPayServer.csproj" />
   </ItemGroup>
+
+  <ItemGroup>
+    <None Remove="uploadtestfile.txt" />
```
Comment (NicolasDorier) [inline BTCPayServer.Tests/BTCPayServer.Tests.csproj]: just create dummy file dynamically

**Comment by Kukks** [comment]
Added tests

**Comment by NicolasDorier** [comment]
Need rebase, then I think we can merge (will need to setup CI later for those dependencies though)

File: BTCPayServer.Tests/StorageTests.cs [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+                var controller = tester.PayTester.GetController<ServerController>(user.UserId, user.StoreId);
+
+//                //For some reason, the tests cache something on circleci and this is set by default
+//                //Initially, there is no configuration, make sure we display the choices available to configure
+//                Assert.IsType<StorageSettings>(Assert.IsType<ViewResult>(await controller.Storage()).Model);
+//
+//                //the file list should tell us it's not configured:
+//                var viewFilesViewModelInitial =
+//                    Assert.IsType<ViewFilesViewModel>(Assert.IsType<ViewResult>(await controller.Files()).Model);
+//                Assert.False(viewFilesViewModelInitial.StorageConfigured);
```
Comment (NicolasDorier) [inline BTCPayServer.Tests/StorageTests.cs]: `For some reason, the tests cache something on circleci and this is set by default` not a problem of circleCI.

The problem is that the tests all run on a shared instance that may have data from previous test.

Comment (NicolasDorier) [reply BTCPayServer.Tests/StorageTests.cs]: Also where does it works? The new account you create should not be admin, so you should not be able to access ServerController stuff.

Comment (Kukks) [reply BTCPayServer.Tests/StorageTests.cs]: resolved: we use controller directly in tests

**Comment by Kukks** [comment]
I've disabled the google/amazon code until someone has time to test it further

**Pull request merged**